prov/gni/include/gnix_mbox_allocator.h

```diff
@@ -0,0 +1,145 @@
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef _GNIX_MBOX_ALLOCATOR_
+#define _GNIX_MBOX_ALLOCATOR_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <stddef.h>
+
+#include "gni_pub.h"
+
+#include "gnix.h"
+#include "gnix_bitmap.h"
+
+enum gnix_page_size {
+	GNIX_PAGE_2MB   = 2,
+	GNIX_PAGE_4MB   = 4,
+	GNIX_PAGE_8MB   = 8,
+	GNIX_PAGE_16MB  = 16,
+	GNIX_PAGE_32MB  = 32,
+	GNIX_PAGE_64MB  = 64,
+	GNIX_PAGE_128MB = 128,
+	GNIX_PAGE_256MB = 256,
+	GNIX_PAGE_512MB = 512
+};
+
+/**
+ * Structure representing mailbox allocated by mbox_allocator.
+ *
+ * @var memory_handle	Memory handle returned by GNI_MemRegister.
+ * @var base		Pointer to the start of the memory returned by mmap.
+ * @var offset		Offset from the base pointer where mailbox is located.
+ */
+struct gnix_mbox {
+	gni_mem_handle_t *memory_handle;
+
+	void *base;
+	ptrdiff_t offset;
+};
+
+/**
+ * Structure representing a slab of memory allocated by the mbox_allocator.
+ *
+ * @var base		The pointer to the start of memory returned by mmap.
+ * @var used		Bitmap of slab usage.
+ * @var memory_handle	The handle returned by GNI_MemRegister.
+ * @var list_entry	Entry for storing structure in slist structs.
+ */
+struct gnix_slab {
+	void *base;
+
+	gnix_bitmap_t used;
+	gni_mem_handle_t memory_handle;
+
+	struct slist_entry *list_entry;
+};
+
+/**
+ * Structure representing an mbox_allocator.
+ *
+ * @var nic_handle	Gnix_nic that the memory is associated with.
+ * @var cq_handle	CQ handle that the memory is associated with.
+ * @var page_size	The page size used for HUGETLBFS and mmap.
+ * @var mbox_size	Size per each mailbox.
+ * @var mpmmap		Mailboxes per mmap slab.
+ */
+struct gnix_mbox_alloc_handle {
+	struct gnix_nic *nic_handle;
+	gni_cq_handle_t cq_handle;
+
+	enum gnix_page_size page_size;
+	size_t mbox_size;
+	size_t mpmmap;
+
+	struct slist slab_list;
+};
+
+/**
+ * Creates an allocator that can be used with mbox_alloc to allocate mailboxes.
+ *
+ * @param nic		IN	Gnix_nic memory is associated with.
+ * @param cq_handle	IN	CQ handle memory is associated with.
+ * @param page_size	IN	Page size used for HUGETLBFS and mmap.
+ * @param mbox_size	IN	Size per each mailbox.
+ * @param mpmmap	IN	Mailboxes per mmap slab.
+ * @param alloc_handle	IN/OUT	Handle needed for use with mbox_alloc.
+ */
+int mbox_allocator_create(struct gnix_nic *nic, gni_cq_handle_t cq_handle,
+			  enum gnix_page_size page_size, size_t mbox_size,
+			  size_t mpmmap,
+			  struct gnix_mbox_alloc_handle *alloc_handle);
+
+/**
+ * Allocate a new mailbox.
+ *
+ * @param alloc_handle	IN	Gnix_mbox_alloc_handle to use as allocator.
+ * @param ptr		IN/OUT	Pointer that contains allocated gnix_mbox.
+ */
+int mbox_alloc(struct gnix_mbox_alloc_handle *alloc_handle,
```

Comment:
Shouldn't it be "struct gnix_mbox *ptr"?  The interface fills out a descriptor allocated by the user.